pkg/cluster/updater.go

```diff
@@ -0,0 +1,372 @@
+	if err != nil {
+		return err
+	}
+	// Update the image in the specific container we care about (should match name
+	// of the deployment itself, per convention).
+	dep.Labels["version"] = image.Tag
+	dep.Spec.Template.Labels["version"] = image.Tag
+	for i, c := range dep.Spec.Template.Spec.Containers {
+		if c.Name == name {
+			dep.Spec.Template.Spec.Containers[i].Image = image.String()
+			break
+		}
+	}
+	// Update the deployment, which will trigger an update.
+	_, err = client.Extensions().Deployments(api.NamespaceSystem).Update(dep)
+	if err != nil {
+		return err
+	}
+	// Wait for update to finish.
+	l, err := labels.Parse(fmt.Sprintf("version=%s,k8s-app=%s", image.Tag, name))
+	if err != nil {
+		return err
+	}
+	lo := api.ListOptions{
+		LabelSelector: l,
+	}
+	replicas := int32(1)
+	if dep.Spec.Replicas != nil {
+		replicas = *dep.Spec.Replicas
+	}
+	for {
+		glog.Info("checking that all new containers are up")
+		upl, err := client.Core().Pods(api.NamespaceSystem).List(lo)
+		if err != nil {
+			// TODO might be worth returning error here.
+			glog.Error(err)
+		} else {
+			if int32(len(upl.Items)) == replicas {
+				if allContainersRunning(upl) {
+					return nil
+				}
+			}
+			glog.Infof("not all new containers have started, expected %d got %d", dep.Status.Replicas, len(upl.Items))
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+func allContainersRunning(pl *v1.PodList) bool {
+	for _, p := range pl.Items {
+		for _, cs := range p.Status.ContainerStatuses {
+			if cs.State.Running == nil {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// OnHostKubeletRollingUpdate will perform a safe rolling update of the on-host kubelets
+// for every node. It does this by placing annotations on each node, which the node-agent
+// should react to, and then waits for the "current-config" annotation to be updated for
+// each node before moving on.
+func OnHostKubeletRollingUpdate(client clientset.Interface, _ string, image *ContainerImage) error {
+	// foreach(node in nodes)
+	// -- updateAnnotation(node)
+	// -- waitForUpdatedCurrentAnnotation(node, timeout)
+	var lo api.ListOptions
+	nl, err := client.Core().Nodes().List(lo)
+	if err != nil {
+		return err
+	}
+	for _, n := range nl.Items {
+		if n.Annotations == nil {
+			n.Annotations = make(map[string]string)
+		}
+		currentConfig := make(map[string]string)
+		err = json.Unmarshal([]byte(n.Annotations[node.CurrentConfigAnnotation]), &currentConfig)
+		if err != nil {
+			return err
+		}
+		currentConfig[node.KubeletVersionKey] = image.Tag
+		b, err := json.Marshal(currentConfig)
+		if err != nil {
+			return err
+		}
+		n.Annotations[node.DesiredConfigAnnotation] = string(b)
+		_, err = client.Core().Nodes().Update(&n)
+		if err != nil {
+			return err
+		}
+		lo := api.ListOptions{
+			Watch:         true,
+			FieldSelector: fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", n.Name)),
+		}
+		w, err := client.Core().Nodes().Watch(lo)
```

Comment:
You may need to establish this watch prior to the update operation, otherwise there is a window for you to miss the node update. But... nodes objects are frequently updated (kubelet will update nodeStatus so we know node is still alive) - where we would catch the change. So if we're okay for waiting for that next window, this may be a non-issue.

So choice is - establish watch prior to making the update
or, rely on the node status update window (should check how long this is) -- but that could artificially delay this process.